### ISSUE TITLE
Create stateful instances for managed instance group

### ISSUE THREAD

**Issue by marko7460** [terraform-providers/terraform-provider-google#5665]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

GCP has an ability to create stateful managed instance groups as described in https://cloud.google.com/compute/docs/instance-groups/stateful-migs. One of the properties of the stateful managed instance groups is ability to preserve disk and instance names.
The way this work using gcloud is:
```bash
$ gcloud compute instance-groups managed create my-test-group --template stateful-instances-template --region us-central1 --instance-redistribution-type NONE --size 0
$ gcloud compute instance-groups managed create-instance my-test-group --instance my-instance-name-1 --region us-central1
$ gcloud compute instance-groups managed create-instance my-test-group --instance my-instance-name-2 --region us-central1
```
This flow will created managed instance group with user defined instance names

At the moment `google_compute_region_instance_group_manager` resource creates instances based on `base_instance_name`, however the instance names have a random string appended at the end.

Having user defined instance names can be useful in the following scenario:
Database cluster of three instances where instance names are mongo1, mongo2, mongo3. All three instances are part of a stateful managed instance group. If we can control names then we can leverage internal DNS to configure each instance at the startup. Furthermore a user can force mongo1 to be a primary database based on it's hostname.

### New or Affected Resource(s)

* google_compute_region_instance_group_manager
* google_compute_instance_group_manager

### Potential Terraform Configuration

```tf
resource "google_compute_region_instance_group_instance" {
    project = var.project
    name = "monog1"
    instance_group = google_compute_region_instance_group_manager.main.self_link
    region = var.region
}
```

### References

* https://cloud.google.com/compute/docs/instance-groups/creating-groups-of-managed-instances#stateful
* https://cloud.google.com/compute/docs/instance-groups/configuring-stateful-migs
* https://github.com/terraform-providers/terraform-provider-google/issues/5635
* https://github.com/terraform-providers/terraform-provider-google/issues/5636

**Comment by kustodian** [comment]
We are very interested in this feature as well, so that we can create [stateful MIGs](https://cloud.google.com/compute/docs/instance-groups/stateful-migs).

**Comment by vDMG** [comment]
We are also very interested on this feature to create our zookeeper cluster on `RECREATE` mode.

**Comment by engrravijain** [comment]
@rileykarson, Any progress on this?

**Comment by rileykarson** [comment]
The issue has been assigned to a sprint starting today, so it should end up with an owner in the next couple of hours.

**Comment by slevenick** [comment]
I'm actively working on this feature, and I have some questions on how this would be used for anyone who was looking to use stateful instances:

The current way I'm planning on implementing this would be a separate resource for stateful instances within an instance group, similar to the sample config in this issue.

1. Do you expect to manage stateful *and* non-stateful instances with the same instance group manager? 

2. Do you expect to configure minimal and most disruptive action allowed for updates to stateful instances individually? 
For example: instance-a and instance-b are stateful instances managed by instance-group-manager-1. Do you expect to be able to configure an update to instance-a to replace the instance, while an update to instance-b can only ever restart the instance?

**Comment by kustodian** [comment]
We would only plan to use these for DBs, so answers to your questions from our point of view would be:

1. No
2. TBH, I have no idea. My gut tells me that if I have 3 instances of e.g. Mongo that we would want to have the same minimal action configured for all of them, since we would like to be able to update all of them when needed, but I guess being able to configure it per instance could probably fill some use case I cannot think of right now. Do you have a specific use case with a concrete example?

**Comment by slevenick** [comment]
This will be available in two parts: 
1. the `stateful_disk` block on https://www.terraform.io/docs/providers/google/r/compute_instance_group_manager.html (and regional version)

2. Two new resources, `google_compute_per_instance_config` that allows specification of individual stateful instances belonging to an instance group manager (and the regional version). These will all be available by release 3.22.0, or partially available before that due to the release cycle.
